File: examples/qualifiers/kotlin/core/array.kt
package salvo.core.array

import salvo.*
import salvo.core.iterator.*
import salvo.core.list.*

fun<T> iter(array: Array<T>): ArrayYield<T> {
    return ArrayYield(items = array, at = 0)
}

data class ArrayYield<T>(
    var items: Array<T>,
    var at: Int,
)

fun<T> next(p: ArrayYield<T>): Union2<T, Finished> {
    val elem = p.items.getOrNull(p.at)
    if (elem == null) {
        return U2_2<T, Finished>(finished())
    }
    p.at = p.at + 1
    return U2_1<T, Finished>(emitted(elem))
}

File: examples/qualifiers/rust/core/array.rs
use crate::core_iterator::*;
use crate::core_list::*;
use crate::unions::*;

pub fn iter<T: Clone + 'static>(array: Vec<T>) -> ArrayYield<T> {
    return ArrayYield { items: array, at: 0 };
}

#[derive(Clone, Debug)]
pub struct ArrayYield<T: Clone + 'static> {
    pub items: Vec<T>,
    pub at: i32,
}

pub fn next<T: Clone + 'static>(p: &mut ArrayYield<T>) -> Union2<T, Finished> {
    let mut elem = p.items.get((p.at) as usize).cloned();
    if elem.is_none() {
        return Union2::<T, Finished>::U2(finished());
    }
    p.at = p.at + 1;
    return Union2::<T, Finished>::U1(emitted(elem.as_ref().unwrap().clone()));
}
